### DISCUSSION TITLE
Composant réseau logique

### DISCUSSION THREAD

**Discussion by mnotin** [dbarzin/mercator#733] [Q&A] [answered]
Bonjour,

Est-il prévu d'ajouter un objet "Périphérique" et "Téléphone" dans la vue "Infrastructure logique" ?
L'objectif ici serait de pouvoir associer un périphérique physique et un téléphone physique à un équivalent logique pour leur associer une adresse IP et ainsi les rattacher à un sous-réseau.
Cela pourrait aussi être réalisé par un objet générique "Composant réseau logique" par exemple, qui ferait ensuite le lien entre un sous-réseau et un objet réseau physique (périphérique, téléphone, etc).

Cordialement

**Comment by dbarzin** [comment]
Bonjour,
Très bonne idée, cela va être ajouté à la feuille de route !
Merci,

**Comment by dbarzin** [comment]
Je me rends compte que pour certains objets, comme les postes de travail, les informations que l'on pourrait retrouver dans l'infrastructure logique se trouvent dans l'infrastructure physique (ex. : adresse IP, OS...). 
Si un objet physique a toujours un objet logique associé, est-il nécessaire de les séparer en deux objets distincts ?

**Comment by mnotin** [comment]
Effectivement, je pense qu'il est dans ce cas peut-être préférable de retirer ces champs de l'objet physique, et de les transférer dans un équivalent logique de l'objet en question.
Il y a deux possibilités :
- soit créer un équivalent logique pour chaque objet physique possédant une adresse IP (ici notamment téléphone, périphérique, poste de travail et éventuellement infrastructure de stockage)
- soit créer un objet logique général (par exemple "Composant réseau logique") qui permettrait de faire le lien entre un objet physique et un sous-réseau.

La première solution est peut-être préférable car tous les objets logiques n'auront pas forcément besoin des mêmes attributs, cela implique cependant la création d'un plus grand nombre d'objets.
Peu importe la solution retenue, le champ adresse IP de l'objet logique permettrait ainsi de faire le lien entre un objet physique et les sous-réseaux auxquels il est rattaché. Ces liens seraient ensuite visibles sur la page d'exploration de la cartographie, cela est par exemple déjà fonctionnel avec les routeurs.

**Comment by sbrt-cy** [comment]
Bonjour,

Merci pour le travail déjà effectué sur cet outil.
Avez-vous une idée de la date de réalisation de cette amélioration ?

Nous souhaitons mettre en œuvre Mercator sur nos infrastructures industrielles composées d'équipements spécifiques auxquels sont rattachées des adresses IP (Caméra, Téléphone, Automates industriels …). Nous devons donc représenter ces équipements sur la vue logique au niveau de la cartographie.

L'ajout de cette fonctionnalité nous permettrait de grandement avancer dans ce sens.

Merci.

**Comment by dbarzin** [comment]
C'est sur la liste des améliorations. 
Ce sera disponible au plus tard début septembre.

**Comment by dbarzin** [comment]
Les objets physiques concernés par cette amélioration de la cartographie incluent :

- Postes de travail
- Infrastructures de stockage
- Périphériques
- Téléphones
- Bornes WiFi
- Éléments de sécurité physique

Trois options sont envisageables pour cette amélioration :

1. **Ajouter un objet logique pour chaque objet physique :**
   - Avantages : Respecte la séparation entre les objets physiques et logiques.
   - Inconvénients : Complexifie le modèle. Cela nécessiterait la création d'un objet logique pour chaque objet physique, ce qui ajouterait 6 nouveaux objets logiques aux 14 déjà existants.

2. **Lier l'objet physique à un serveur logique :**
   - Avantages : Respecte également la séparation physique/logique.
   - Inconvénients : Implique la déclaration de deux objets distincts. Cela peut également entraîner des incohérences, car les serveurs logiques contiendraient des objets de type téléphone, borne WiFi, etc., ce qui n'est pas souhaitable.

3. **Ajouter les champs nécessaires (IP, OS, CPE, etc.) directement dans l'objet physique :**
   - Avantages : Solution la plus simple et la plus directe à implémenter.
   - Inconvénients : Ne respecte pas la division stricte entre les objets physiques et logiques.

Pour le moment, je penche en faveur de la troisième solution, qui semble plus pragmatique. Elle simplifie l'implémentation tout en répondant aux besoins immédiats.

Qu'en pensez-vous ?

**Reply by seb976** [reply]
Je pense effectivement que le mieux, serait de rajouter les différents champs liés à l'objet logique dans l'objet physique.
Dans l'idéal, il faudrait délier les objets des vues dans le menu déroulant. Crée une section objet général, et que dans chaque objet, il y ait une partie configuration physique et une partie configuration logique.

**Comment by mnotin** [comment]
Merci beaucoup pour votre retour.

Notre besoin principal par rapport à cette fonctionnalité serait de pouvoir créer un lien visuel entre un objet physique et un sous-réseau. Au niveau de la vue d'exploration notamment et éventuellement au niveau de la vue de l'infrastructure logique si la représentation des ces objets est ajoutée dans cette vue également.

En ce sens, la solution 1 pourrait convenir, il s'agirait d'étendre le fonctionnement actuel des routeurs en pouvant faire un lien entre un objet physique et un objet logique puis entre un objet logique et un sous-réseau.

La solution 2 me semble peu enviable au vu des différents inconvénients que vous avez soulevés.

La solution 3 pourrait fonctionner en ajoutant un lien entre un objet physique et un sous-réseau visible depuis la vue d'exploration, en ajoutant un attribut "Adresse IP" aux objets physiques concernés. En plus de ce lien, un représentation de l'objet physique pourrait être réalisée également dans la vue logique, mais seulement à titre indicatif, afin de pouvoir également visualiser ce lien depuis la vue de l'infrastructure logique.

Une quatrième solution pourrait être envisagée, qui serait basée sur la création d'un objet logique qui permettrait de faire le lien entre un objet physique et un sous-réseau.
Cet objet pourrait s'appeler "Lien réseau" ou "Lien logique" par exemple et contenir deux champs :
- Objet physique : permettrait de sélectionner un objet physique parmi tous ceux pouvant contenir une adresse IP, c'est-à-dire :
  - Serveur physique
  - Poste de travail
  - Infrastructure de stockage
  - Périphérique
  - Téléphone
  - Commutateur physique
  - Routeur physique
  - Borne Wi-Fi
  - Equipement de sécurité physique
- Adresse(s) IP : permettrait à Mercator de faire le lien entre un objet physique et un ou plusieurs sous-réseau(x) associé(s).

L'avantage ici est que cette solution limite le nombre d'objets logiques et permet de garder la division logique / physique.

En synthèse, la solution 4 passant par l'utilisation d'un objet logique générique serait la plus appropriée dans notre cas.
Autrement, la solution 1 ou 3 avec l'amélioration proposée pourrait répondre à notre besoin.

**Comment by dbarzin** [comment]
:tada: La solution  3 est en place pour les périphériques et les postes de travail ! https://github.com/dbarzin/mercator/pull/800

**Comment by mnotin** [comment]
Merci beaucoup pour cette amélioration !
Un ajout futur des autres objets mentionnés précédemment est-il toujours prévu ?

J'en profite également pour vous remonter une problématique que je rencontre après avoir mis à jour Mercator sur cette dernière version : il ne m'est plus possible d'ajouter d'objet dans la vue d'exploration de la cartographie, je reçois une erreur `ReferenceError: addNode is not defined at HTMLAnchorElement.onclick` dans la console.
Pour information ce problème ne se produit pas lorsque je réalise une nouvelle installation propre de Mercator, j'ai donc peut-être fait une erreur lors de la mise à jour.

**Reply by dbarzin** [reply]
L'ajout pour les autres objets est prévu dans les prochains jours.
Peux-tu poster une issue pour le problème d'exploration de la cartographie ?

**Reply by mnotin** [reply]
D'accord très bien, je vous remercie.
Concernant le problème d'exploration, il a été réglé grâce aux modifications que vous avez apportées ce dimanche 25, merci à vous.

**Comment by dbarzin** [comment, marked as answer]
🎉 La solution 3 est en place pour les autres objets ! https://github.com/dbarzin/mercator/pull/804

**Comment by mnotin** [comment]
Merci beaucoup pour l'ajout de cette fonctionnalité !
